## Introduction_to_Neural_Networks_and_Deep_Learning_R7_Project2


In [0]:

import pandas as pd
import h5py
import numpy as np

In [0]:
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [0]:
images = h5py.File('/content/gdrive/My Drive/AIML/DeepLearning/SVHN_single_grey1.h5', 'r')

In [0]:
X_train = images['X_train'][:]
X_test = images['X_test'][:]
X_val = images['X_val'][:]
y_train = images['y_train'][:]
y_test = images['y_test'][:]
y_val = images['y_val'][:]

X_test = np.array(X_test)
X_train = np.array(X_train)
X_val = np.array(X_val)
y_test = np.array(y_test)
y_train = np.array(y_train )
y_val = np.array(y_val)



In [0]:
# Implementing  the KNN model
from sklearn.neighbors import KNeighborsClassifier 
from imutils,cv2,argparse,os
import matplotlib.pyplot as plt
%matplotlib inline

In [0]:
# Reshaping the images
X_train_reshaped = np.reshape(X_train, (X_train.shape[0], -1))
X_test_reshaped = np.reshape(X_test, (X_test.shape[0], -1))
X_val_reshaped = np.reshape(X_val, (X_val.shape[0], -1))

In [0]:
knn = KNeighborsClassifier(n_neighbors = 10)

In [0]:
knn.fit(X_train_reshaped, trainY)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=10, p=2,
                     weights='uniform')

In [0]:
print("Accuracy with k = 10: ", knn.score(X_val_reshaped, valY)*100, '%')

Accuracy with k = 10:  54.1 %


In [0]:
# Implementing DNN model


In [0]:
trainY = tf.keras.utils.to_categorical(y_train, num_classes=10)
testY = tf.keras.utils.to_categorical(y_test, num_classes=10)
valY = tf.keras.utils.to_categorical(y_val, num_classes=10)

In [2]:
import tensorflow as tf
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Reshape((1024,),input_shape=(32,32,)))
model.add(tf.keras.layers.BatchNormalization())

Instructions for updating:
If using Keras pass *_constraint arguments to layers.


In [0]:
#Dense layers
model.add(tf.keras.layers.Dense(50, activation='relu'))
model.add(tf.keras.layers.Dense(100, activation='relu'))
model.add(tf.keras.layers.Dense(200, activation='relu'))
model.add(tf.keras.layers.Dense(100, activation='relu'))
model.add(tf.keras.layers.Dense(50, activation='relu'))

In [0]:
#Final output layer
model.add(tf.keras.layers.Dense(10, activation='softmax'))

In [0]:

model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [8]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape (Reshape)            (None, 1024)              0         
_________________________________________________________________
batch_normalization (BatchNo (None, 1024)              4096      
_________________________________________________________________
dense (Dense)                (None, 200)               205000    
_________________________________________________________________
dense_1 (Dense)              (None, 100)               20100     
_________________________________________________________________
dense_2 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_3 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_4 (Dense)              (None, 100)               5

In [0]:
model.fit(X_train,trainY,          
          validation_data=(X_test,testY),
          epochs=10,
          batch_size=64)

Train on 42000 samples, validate on 18000 samples
Epoch 1/10
42000/42000 [==============================] - 4s 95us/sample - loss: 1.9493 - acc: 0.3119 - val_loss: 2.0793 - val_acc: 0.3408
Epoch 2/10
42000/42000 [==============================] - 4s 84us/sample - loss: 1.2459 - acc: 0.6070 - val_loss: 1.3378 - val_acc: 0.5779
Epoch 3/10
42000/42000 [==============================] - 4s 84us/sample - loss: 1.0054 - acc: 0.6894 - val_loss: 1.0966 - val_acc: 0.6656
Epoch 4/10
42000/42000 [==============================] - 4s 86us/sample - loss: 0.8797 - acc: 0.7276 - val_loss: 1.1077 - val_acc: 0.6661
Epoch 5/10
42000/42000 [==============================] - 4s 84us/sample - loss: 0.7994 - acc: 0.7527 - val_loss: 1.3135 - val_acc: 0.6175
Epoch 6/10
42000/42000 [==============================] - 4s 85us/sample - loss: 0.7405 - acc: 0.7716 - val_loss: 1.2088 - val_acc: 0.6560
Epoch 7/10
42000/42000 [==============================] - 4s 86us/sample - loss: 0.6939 - acc: 0.7851 - val_loss: 0.

In [0]:
#Checking for 10 more iterations
model.fit(X_train,trainY,validation_data=(X_test,testY),epochs=10,batch_size=64)

Train on 42000 samples, validate on 18000 samples
Epoch 1/10
42000/42000 [==============================] - 4s 86us/sample - loss: 0.4026 - acc: 0.8760 - val_loss: 0.7210 - val_acc: 0.7979
Epoch 2/10
42000/42000 [==============================] - 3s 82us/sample - loss: 0.3966 - acc: 0.8780 - val_loss: 0.7671 - val_acc: 0.7874
Epoch 3/10
42000/42000 [==============================] - 3s 82us/sample - loss: 0.3943 - acc: 0.8784 - val_loss: 0.8376 - val_acc: 0.7717
Epoch 4/10
42000/42000 [==============================] - 3s 83us/sample - loss: 0.3876 - acc: 0.8804 - val_loss: 0.7107 - val_acc: 0.8002
Epoch 5/10
42000/42000 [==============================] - 3s 81us/sample - loss: 0.3799 - acc: 0.8826 - val_loss: 0.6355 - val_acc: 0.8253
Epoch 6/10
42000/42000 [==============================] - 3s 82us/sample - loss: 0.3725 - acc: 0.8850 - val_loss: 0.6490 - val_acc: 0.8268
Epoch 7/10
42000/42000 [==============================] - 3s 83us/sample - loss: 0.3687 - acc: 0.8856 - val_loss: 0.

In [0]:
#Checking for 10 more iterations

model.fit(X_train,trainY,          
          validation_data=(X_test,testY),
          epochs=10,
          batch_size=64)

Train on 42000 samples, validate on 18000 samples
Epoch 1/10
42000/42000 [==============================] - 4s 98us/sample - loss: 0.3421 - acc: 0.8910 - val_loss: 0.5966 - val_acc: 0.8426
Epoch 2/10
42000/42000 [==============================] - 4s 92us/sample - loss: 0.2983 - acc: 0.9057 - val_loss: 0.6119 - val_acc: 0.8400
Epoch 3/10
42000/42000 [==============================] - 4s 88us/sample - loss: 0.2907 - acc: 0.9067 - val_loss: 0.6040 - val_acc: 0.8469
Epoch 4/10
42000/42000 [==============================] - 4s 91us/sample - loss: 0.2832 - acc: 0.9086 - val_loss: 0.5947 - val_acc: 0.8473
Epoch 5/10
42000/42000 [==============================] - 4s 87us/sample - loss: 0.2768 - acc: 0.9117 - val_loss: 0.6141 - val_acc: 0.8469
Epoch 6/10
42000/42000 [==============================] - 4s 90us/sample - loss: 0.2714 - acc: 0.9123 - val_loss: 0.6428 - val_acc: 0.8370
Epoch 7/10
42000/42000 [==============================] - 4s 87us/sample - loss: 0.2716 - acc: 0.9125 - val_loss: 0.

# Conclusion


**KNN Model gave very less accuracy i.e 54% and took very long time for computation . Deep Neural Network model gave very good accuracy of 92% with considerably less time when compared to KNN model**
---

